src/finalfusion/subword/hash_indexers.pyx

```diff
@@ -0,0 +1,325 @@
+        ------
+        TypeError
+            If `word` is None.
+        """
+        if word is None:
+            raise TypeError("Can't extract ngrams for None")
+        cdef unsigned int start, end
+        cdef Py_ssize_t i, j
+        cdef uint64_t h
+        if bracket:
+            word = "<%s>" % word
+        cdef bytes b_word = word.encode("utf-8")
+        cdef const unsigned char*b_word_ptr = b_word
+        cdef Py_ssize_t length = len(word)
+        cdef uint32_t max_n = min(self._max_n, length)
+        cdef array.array offsets = find_utf8_boundaries(b_word_ptr, len(b_word))
+        cdef list ngrams = []
+        # iterate over starting points by character
+        for i in range(length + 1 - self._min_n):
+            # offsets[i] corresponds to the byte offset to the start of the char at word[i]
+            start = offsets.data.as_uints[i]
+            # iterate over ngram lengths, long to short
+            for j in range(max_n, self._min_n - 1, -1):
+                if j + i <= length:
+                    # offsets[i+j] holds the exclusive upper bounds of the bytes for the character word[i+j-1]
+                    end = offsets.data.as_uints[i + j]
+                    h = ft_hash_ngram(b_word_ptr, start, end) % self._n_buckets + offset
+                    if with_ngrams:
+                        ngrams.append((h, word[i:i + j]))
+                    else:
+                        ngrams.append(h)
+        return ngrams
+
+    @property
+    def idx_bound(self) -> int:
+        return self._n_buckets
+
+    def __repr__(self):
+        return f'FastTextIndexer(n_buckets={self._n_buckets}, min_n={self._min_n}, max_n={self._max_n})'
+
+cdef array.array find_utf8_boundaries(const unsigned char*w, const Py_ssize_t n_bytes):
+    cdef unsigned int b
+    cdef Py_ssize_t i = 0
+    cdef array.array offsets = array.array('I')
+    # n_bytes + 1 to store n_bytes as final boundary
+    array.resize(offsets, n_bytes + 1)
+    for b in range(n_bytes):
+        # byte w[b] is not a continuation byte, therefore beginning of char; store offset
+        if (w[b] & 0xC0) != 0x80:
+            offsets.data.as_uints[i] = b
+            i += 1
+    offsets.data.as_uints[i] = <unsigned int> n_bytes
+    return offsets
+
+cdef uint64_t SEED64 = 0xcbf29ce484222325
+cdef uint64_t PRIME64 = 0x100000001b3
+cdef uint32_t SEED32 = 2166136261
+cdef uint32_t PRIME32 = 16777619
+
+cdef uint64_t fifu_hash_ngram(str word, const Py_ssize_t start, const Py_ssize_t length):
+    cdef uint32_t c
+    cdef Py_ssize_t i = 0
+    cdef uint64_t h = SEED64
+
+    h = fnv64(<uint8_t*> &length, 8, h=h)
+    for i in range(start, start + length):
+        # extract unicode for char, cast to u8 pointer for hashing.
+        # extracting bytes manually from memoryview/bytes is
+        # more complex because of handling prefixes.
+        c = ord(word[i])
+        h = fnv64(<uint8_t*> &c, 4, h=h)
+    return h
+
+cdef uint64_t fnv64(const uint8_t*data,
```

Comment:
If you hand-roll an implementation, test against the provided test vectors:

http://www.isthe.com/chongo/src/fnv/test_fnv.c